### PR TITLE
ddl: use static contexts in `NewReorgCopContext`

### LINKED ISSUE #53388: remove `mock.Context` usage in production code
## Enhancement

It's better to remove the `mock.Context` usages in TiDB production code. The reason is that `mock.Context` is not designed for production code. It is designed for testing, that means some values in it are not reasonable in production and some values will be modified by the developer without any care.

We can replace some code that uses `mock.Context` with some "static" context we introduced before such as `StaticExprContext` or `DistSQLContext`

## Tasks

- #53378
- #53389
- #53441
- #53467
- #53479
- #53642
- #53671
- #53721
- #55823
- #55873
- #56038
- #56174
- #56176
- #56348
- #56355
- #56654

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: ref #53388

### What changed and how does it work?

This PR:

- Introduces new functions `newReorgExprCtxWithReorgMeta` and `newReorgDistSQLCtxWithReorgMeta` to build `ExprContext` and `DistSQLContext` for DDL reorg.
- Use above methods to construct static contexts in `NewReorgCopContext`

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/util/mock/context.go [hunk 1]
```diff
@@ -272,6 +272,7 @@ func (c *Context) GetDistSQLCtx() *distsqlctx.DistSQLContext {
 		TiFlashMaxBytesBeforeExternalSort:    vars.TiFlashMaxBytesBeforeExternalSort,
 		TiFlashMaxQueryMemoryPerNode:         vars.TiFlashMaxQueryMemoryPerNode,
 		TiFlashQuerySpillRatio:               vars.TiFlashQuerySpillRatio,
+		ResourceGroupName:                    sc.ResourceGroupName,
```
Comment (lcwangchao) [inline pkg/util/mock/context.go:275]: It seems `ResourceGroupName` is lost in the previous implementation.  PTAL @YangKeao

File: pkg/util/mock/context.go [hunk 2]
```diff
@@ -641,8 +642,8 @@ func NewContext() *Context {
 	vars.MemTracker.SetBytesLimit(-1)
 	vars.DiskTracker.SetBytesLimit(-1)
 	vars.StmtCtx.MemTracker, vars.StmtCtx.DiskTracker = memory.NewTracker(-1, -1), disk.NewTracker(-1, -1)
-	vars.MemTracker.AttachTo(vars.MemTracker)
-	vars.DiskTracker.AttachTo(vars.DiskTracker)
+	vars.StmtCtx.MemTracker.AttachTo(vars.MemTracker)
```
Comment (lcwangchao) [inline pkg/util/mock/context.go:645]: After discussing with @XuHuaiyu , we should use `vars.StmtCtx.MemTracker` here

**Review by tangenta — APPROVED** [review]

**Review by YangKeao — APPROVED** [review]
LGTM

**Pull request merged**